[docs] #39, #83, #90: Update WASM/FFI guide (#114)
hyperledger/iroha-2-docs: src/guide/advanced/ffi.md

# Foreign Function Interfaces (FFI)

As we mentioned before, Iroha 2 relies on the C-linkage to generate
[WASM](./wasm.md) bindings, and the `iroha_wasm` crate deals with foreign

> function interfaces (FFI). Here we discuss in more details how Iroha 2

Reviewer: ```suggestion
function interfaces (FFI). Here we will discuss in more detail how Iroha 2
```

handles FFI logic.

## FFI Binding Generation

The `iroha_ffi` crate is used to generate functions that are callable via
FFI.

A Rust structure is converted into an intermediary FFI struct via `IntoFfi`

> trait, which implements the `AsReprCRef` trait and can be converted into a

Reviewer: ```suggestion
trait, which implements the `AsReprCRef` trait, and can be converted into a
```

type that can cross FFI boundary. This goes the other way around as well:
FFI ReprC type is converted into a struct via `TryFromReprC`.

The diagram below uses the creation of a new domain as an example to show
the conversion process (more on the name mangling semantics in a
[separate section](#name-mangling)).

![Untitled](/img/ffi.png)

The traits that enable binding generation are `ReprC`, `AsReprCRef`,
`TryFromReprC`, and `IntoFfi`:

| Trait          | Description                                                                                                       |
| -------------- | ----------------------------------------------------------------------------------------------------------------- |

> | `ReprC`        | The trait represents the robust type that conforms to C ABI. The type can be safely shared across FFI boundaries. |

Reviewer: ```suggestion
| `ReprC`        | This trait represents the robust type that conforms to C ABI. The type can be safely shared across FFI boundaries. |
```

> | `AsReprCRef`   | The trait is used to convert from a Rust type reference to `ReprC` reference.                                     |

Reviewer: ```suggestion
| `AsReprCRef`   | This trait is used to convert from a Rust type reference to `ReprC` reference.                                     |
```

> | `TryFromReprC` | The trait is used to perform conversion from a type that implements `ReprC`.                                      |

Reviewer: ```suggestion
| `TryFromReprC` | This trait is used to perform a conversion from a type that implements `ReprC`.                                      |
```

> | `IntoFfi`      | The trait is used to convert into a type that can be converted to an FFI-compatible `ReprC` type.                 |

Reviewer: ```suggestion
| `IntoFfi`      | This trait is used to convert into a type that can then be converted to an FFI-compatible `ReprC` type.                 |
```

Note that there is no ownership transfer over FFI except for opaque pointer
types.

### Name Mangling

Note the use of double underscores in generated names of FFI objects:

- For the `inherent_fn` function in the `StructName` struct, the FFI name
  would be `StructName__inherent_fn`.
- For the `trait_method_name` method in `StructName` struct, the FFI name

>   would be `StructName__trait_method_name`.

Reviewer 2: Yes

- For setting the `field_name` field in `StructName` struct, the FFI name
  would be `StrucuName__set_field_name`.
- For getting the `field_name` field in `StructName` struct, the FFI name
  would be `StrucuName__field_name`.
- For getting the mutable `field_name` field in `StructName` struct, the
  FFI name would be `StrucuName__field_name_mut`.

> - For the `method_name` method in `module_name`, the FFI name would be
>   `module__name::__method_name`.

Reviewer 3: ```suggestion
- For the freestanding `fn_name` function in `module_name`, the FFI name would be
  `module_name::__fn_name`.
```

there are two things, freestanding functions as above and then there are some traits that are not generic and that allow for sharing of implementation in the FFI
```rs
impl Clone for Type1 {
    fn clone(&self) -> Self;
}
impl Clone for Type2 {
    fn clone(&self) -> Self;
}
```

for which `module_name::__clone` will be generated in the same regard as for freestanding functions

# Foreign Function Interfaces (FFI)

As we mentioned before, Iroha 2 relies on the C-linkage to generate
[WASM](./wasm.md) bindings, and the `iroha_wasm` crate deals with foreign
function interfaces (FFI). Here we will discuss in more detail how Iroha 2
handles FFI logic.

## FFI Binding Generation

The `iroha_ffi` crate is used to generate functions that are callable via
FFI.

> A Rust structure is converted into an intermediary FFI struct via `IntoFfi`

Reviewer: ```suggestion
A Rust structure is converted into an intermediary FFI struct via the `IntoFfi`
```

# Foreign Function Interfaces (FFI)

As we mentioned before, Iroha 2 relies on the C-linkage to generate
[WASM](./wasm.md) bindings, and the `iroha_wasm` crate deals with foreign
function interfaces (FFI). Here we will discuss in more detail how Iroha 2

> handles FFI logic.

Author: applied in #158

## FFI Binding Generation

The `iroha_ffi` crate is used to generate functions that are callable via

> FFI.

Author: applied in #158
